README.md
# Óg [![Build Status](https://travis-ci.org/stephenmoloney/og.svg)](https://travis-ci.org/stephenmoloney/og) [![Hex Version](http://img.shields.io/hexpm/v/og.svg?style=flat)](https://hex.pm/packages/og) [![Hex docs](http://img.shields.io/badge/hex.pm-docs-green.svg?style=flat)](https://hexdocs.pm/og)

[Óg](http://hexdocs.pm/og/Og.html) is a small collection of debugging functions for logging data.


### Note

- Functions `Og.log/2` and `Og.log_r/2` are primarily intended as debugging tools.


### Installation

Add óg to your list of dependencies in `mix.exs`:

```elixir
def deps, do: [{:og, "~> 1.0"}]
```

Ensure that `:logger` is started in the applications:

```elixir
def application do [applications: [:logger, :og]] end
```

## Summary

- `log/2` - logs the data transformed by the inspector function
and returns `:ok`


- `log_r/2` - logs the data transformed by the inspector function
and returns the original data.


- Inspection of the data before logging it can be helpful in a debugging context for
example for
    - Avoiding the `Protocol.UndefinedError` when logging tuples.
    - Not needing to require Logger


- Formatting with `Kernel.inspect` on the data carries an overhead so
`Og.log/2` and `Og.log_r/2` should be probably be reserved for debugging code in `:dev`
environments.


## Configuration

    config :logger, :og,
      kernel_opts: [width: 70],
      apex_opts: [numbers: :false, color: :false],
      sanitize_by_default: :false,
      default_inspector: :kernel


## Configuration options

- `kernel_opts` - corresponds to elixir inspect opts for `IO.inspect/2`
and `Kernel.inspect/2`, refer to `https://hexdocs.pm/elixir/Inspect.Opts.html`


- `apex_opts` - corresponds to options for the `Apex.Format.format/2` function,
refer to `https://github.com/BjRo/apex/blob/master/lib/apex/format.ex`


- `sanitize_by_default` - defaults to `:false`, when set to `:true`, an attempt will
be made to apply the `SecureLogFormatter.sanitize/1` function on the data
before applying the inspection function. For this function to take any
effect, the settings for `SecureLogFormatter` must also be placed in
`config.exs`. See the following
[secure_log_formatter url](https://github.com/localvore-today/secure_log_formatter/blob/master/lib/secure_log_formatter.ex)
for more details.


- `default_inspector` - corresponds to the default inspector which will apply
to the data passed to the log function. This can be overriden in the options
of the log function. The options are `:kernel` or `:apex`, the default is `:kernel`.


Example configuration for secure_log_formatter, as referenced at
the following [secure_log_formatter url](https://github.com/localvore-today/secure_log_formatter/blob/master/lib/secure_log_formatter.ex).


    config :logger,
      secure_log_formatter:
        [
          # Map and Keyword List keys who's value should be hidden
          fields: ["password", "credit_card", ~r/.*_token/],

          # Patterns which if found, should be hidden
          patterns: [~r/4[0-9]{15}/] # Simple credit card example

          # defaults to "[REDACTED]"
          replacement: "[PRIVATE]"
        ]


### some examples


- Basic logging

```elixir
Og.log(:this_is_a_test)
```

- Logging at the `:warn` level

```elixir
Og.log(:this_is_a_test, level: :warn)
```

- Logging at the `:warn` level and with __ENV__ specified to get richer information

```elixir
Og.log(:this_is_a_test, level: :warn, env: __ENV__)
````

- Logging with the Apex inspector

```elixir
Og.log(:this_is_a_test, inspector: :apex)
```

- Logging and applying sanitization with the features of
[secure_log_formatter](https://github.com/localvore-today/secure_log_formatter)

```elixir
Og.log(%{credit_card: 4111111111111}, sanitize: :true)
```

- Logging inside a chain of piped functions

```elixir
defmodule OgTest do
  def log() do
    %{first: "john", last: "doe"}
    |> Map.to_list()
    |> Enum.filter( &(&1 === {:first, "john"}))
    |> Og.log_r()
    |> List.last()
    |> Tuple.to_list()
    |> List.last()
    |> Og.log_r(env: __ENV__, inspector: :kernel, level: :info)
    |> String.upcase()
  end
end
OgTest.log()
```


### Dependencies

- [Apex library](https://hex.pm/packages/apex) --- [Björn Rochel](https://hex.pm/users/bjro)
    - Setting the `config.exs` opts or the log function opts to `inspector: :apex`
    will use the `Apex.Format.format/2` function from the apex library.


- [SecureLogFormatter library](https://hex.pm/packages/secure_log_formatter)  ---  [Sean Callan](https://hex.pm/users/doomspork)
    - Setting `config.exs` opts or the log function opts to `sanitize: :true`
    will use the `SecureLogFormatter.sanitize/1` function from the SecureLogFormatter library.


### Todo

- [ ] Investigate adding a custom formatting module as an optional additional means of logging.


### Licence

MIT


CHANGELOG.md
# Changelog

## v1.1

[changes]
-add `:secure_log_formatter` to applications


## v1.1

[bug fix]
- logger was logging `:warn` as `:error` for some calls.
- `:default_inspector` was absent from configuration - added back in.

[enhancements]
- introduce `:secure_log_formatter` so that sensitive parameters can be
redacted out of the log files. Added `SecureLogFormatter.sanitize(data)`
so that an attempt can optionally be made to sanitize data prior to logging it.

[changes]


## v1.0.2

[changes]
- Add `:apex` to applications. Giving warnings as unstarted application otherwise.

## v1.0.1

[changes]
- Add support for the old form of writing the log functions. Eg `Og.log_r(%{"test" => "test"}, __ENV__, :error)`
as it was faster and quite convenient. These functions will be docless... to avoid cluttering the docs.


## v1.0.0

[security fix]
  - Remove dependency on `Code.eval_string/3`. A potential security issue exists with the use of
[Code.eval_string/3](https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107).
Hence only versions of `Og` >= 1.0.0 should be used.

[Breaking changes]
  - Major change to the api. As the number of arguments grew, the api became complex.
  The api is changing in favour of `log(data, opts)` or `log_r(data, opts)` to simplify the api.



## v0.2.4

***Security Warning: Versions of `Og` less than 1.0.0 are deprecated and should not be used***

- A potential security issue exists with the use of
[Code.eval_string/3](https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)
This was noticed upon reading a relatively new warning in the
[docs]((https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)).
- `Og` versions lower than `v1.0.0` are now deprecated and retired in favour of versions >= `1.0`.

[bug fix]

- Some of the `log_r`, `alog_r` and `klog_r` functions were pointing to the
 incorrect `log`, `alog` and `klog` functions causing the tuple `:ok` to be returned rather than
 the expected data.


## v0.2.3

***Security Warning: Versions of `Og` less than 1.0.0 are deprecated and should not be used***

- A potential security issue exists with the use of
[Code.eval_string/3](https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)
This was noticed upon reading a relatively new warning in the
[docs]((https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)).
- `Og` versions lower than `v1.0.0` are now deprecated and retired in favour of versions >= `1.0`.

[bug fix]

- Misnamed function during refactoring, `logr` -> `log_r`


## v0.2.2

***Security Warning: Versions of `Og` less than 1.0.0 are deprecated and should not be used***

- A potential security issue exists with the use of
[Code.eval_string/3](https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)
This was noticed upon reading a relatively new warning in the
[docs]((https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)).
- `Og` versions lower than `v1.0.0` are now deprecated and retired in favour of versions >= `1.0`.

[bug fix]

- Issue with the default options breaking some the `alog` functions.


## v0.2.1

***Security Warning: Versions of `Og` less than 1.0.0 are deprecated and should not be used***

- A potential security issue exists with the use of
[Code.eval_string/3](https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)
This was noticed upon reading a relatively new warning in the
[docs]((https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)).
- `Og` versions lower than `v1.0.0` are now deprecated and retired in favour of versions >= `1.0`.

[changes]

- Add some docs

## v0.2.0

***Security Warning: Versions of `Og` less than 1.0.0 are deprecated and should not be used***

- A potential security issue exists with the use of
[Code.eval_string/3](https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)
This was noticed upon reading a relatively new warning in the
[docs]((https://github.com/elixir-lang/elixir/commit/f1daca5be78e6a466745ba2cdc66d9787c3cf47f#diff-da151e1c1d9b535259a2385407272c9eR107)).
- `Og` versions lower than `v1.0.0` are now deprecated and retired in favour of versions >= `1.0`.

[changes]
- fix compiler warnings on newer elixir verson.
- Add pull request for `travis.yml`
- remove docs about `log/3`, `log/4` and so on. Add a lot of bulk to the api.
Users can read source code if those functions are needed.

[breaking changes]
- deprecate `context` and `conn_context` functions. Perhaps will create a separate
package for those.
- `log_return` is now renamed to `log_r` for brevity but api will continue to work for
`log_return` for now at least.

[enhancements]
- able to add `inspect_opts` to the config file so that they will be applied by default on
all logs. Example `config :og, inspect_opts: [syntax_colors: [atom: :blue]]`
- add an option to use [Apex](https://github.com/BjRo/apex) for the formatting. See `https://github.com/BjRo/apex`.
- add `klog` function for `Kernel.inspect/2` on data and `alog` function for `Apex.Format.format/2` on data.
- add `klog_r` function for `Kernel.inspect/2` on data and `alog_r` function for `Apex.Format.format/2` on data.
- new dependency on `Apex`.


## v0.1.0

***Security Warning: Versions of `Og` less than 1.0.0 are deprecated and should not be used***

- A potential security issue exists with the use of

config/test.exs
use Mix.Config

config :logger,
  backends: [:console],
  compile_time_purge_level: :info

config :logger, :og,
  kernel_opts: [width: 70],
  apex_opts: [numbers: :false, color: :false],
  sanitize_by_default: :false,
  default_inspector: :kernel

config :logger,
  secure_log_formatter:
    [
      fields: [~r/\w*_token/, "credit_card", "password"],
      patterns: [
        ~r/#?(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})/,
        ~r/\d{3}-?\d{2}-?\d{4}/
      ],
      replacement: "[HIDDEN]"
    ]

test/og_test.exs
defmodule OgTest do
  use ExUnit.Case, async: :false
  alias ExUnit.CaptureLog
  alias Og.TestSupport

  test "log/2 when capture level = :debug" do
    actual = CaptureLog.capture_log(
      [level: :debug],
      fn() -> Og.log("test", level: :info) end
    )
    as_expected? = Regex.match?(~r/test/, actual)
    assert(as_expected? ==  :true)
  end

  test "log/1 when capture level = :warn" do
    actual = CaptureLog.capture_log(
      [level: :warn],
      fn() -> Og.log("test") end
    )
    as_expected? = Regex.match?(~r/test/, actual)
    refute(as_expected? ==  :true)
  end

  test "log/2 when capture level = :debug and log_level = :warn" do
    actual = CaptureLog.capture_log(
      [level: :debug],
      fn() -> Og.log("test", [level: :warn]) end
    )
    as_expected? = Regex.match?(~r/test/, actual)
    assert(as_expected? ==  :true)
  end

  test "log/2 when capture level = :warn and log_level = :warn" do
    actual = CaptureLog.capture_log(
      [level: :warn],
      fn() -> Og.log("test", [level: :warn]) end
    )
    as_expected? = Regex.match?(~r/test/, actual)
    assert(as_expected? ==  :true)
  end

  test "log/2 when capture level = :error and log_level = :warn" do
    actual = CaptureLog.capture_log(
      [level: :error],
      fn() -> Og.log("test", [level: :warn]) end
    )
    as_expected? = Regex.match?(~r/test/, actual)
    refute(as_expected? ==  :true)
  end

  test "log/2 when passing the __ENV__ variable" do
    actual =  TestSupport.log_test(:debug, :info)
    as_expected? = Regex.match?(~r/test/, actual) &&
                              Regex.match?(~r/module: Elixir.Og.TestSupport/, actual) &&
                              Regex.match?(~r/function: log_test\/2/, actual) &&
                              Regex.match?(~r/line: /, actual)
    assert(as_expected? ==  :true)
  end

  test "log/3 when passing the __ENV__ variable and
          when capture level = :debug and log_level = :warn" do
    actual =  TestSupport.log_test(:debug, :warn)
    as_expected? = Regex.match?(~r/test/, actual) &&
                              Regex.match?(~r/module: Elixir.Og.TestSupport/, actual) &&
                              Regex.match?(~r/function: log_test\/2/, actual) &&
                              Regex.match?(~r/line: /, actual)
    assert(as_expected? ==  :true)
  end

  test "log/3 when passing the __ENV__ variable and
          when capture level = :warn and log_level = :warn" do
    actual =  TestSupport.log_test(:warn, :warn)
    as_expected? = Regex.match?(~r/test/, actual) &&
                              Regex.match?(~r/module: Elixir.Og.TestSupport/, actual) &&
                              Regex.match?(~r/function: log_test\/2/, actual) &&
                              Regex.match?(~r/line: /, actual)
    assert(as_expected? ==  :true)
  end

  test "log/3 when passing the __ENV__ variable and
          when capture level = :error and log_level = :warn" do
    actual =  TestSupport.log_test(:error, :warn)
    as_expected? = Regex.match?(~r/test/, actual) &&
                              Regex.match?(~r/module: Elixir.Og.TestSupport/, actual) &&
                              Regex.match?(~r/function: log_test\/2/, actual) &&
                              Regex.match?(~r/line: /, actual)
    refute(as_expected? ==  :true)
  end

  test "log_r/1 logs intermediate data in a pipeline of trasformations" do
    actual = TestSupport.log_test_log_r1()
    as_expected? = Regex.match?(~r/[first: "john"]/, actual)
    assert(as_expected? == :true)
  end

  test "log_r/2 logs intermediate data in a pipeline of trasformations" do
    actual = TestSupport.log_test_log_r2()
    as_expected? = Regex.match?(~r/[first: "john"]/, actual) &&
                              Regex.match?(~r/module: Elixir.Og.TestSupport/, actual) &&
                              Regex.match?(~r/function: log_test_log_r2\/0/, actual) &&
                              Regex.match?(~r/line: /, actual)
    assert(as_expected? == :true)
  end

  test "log_r/3 logs intermediate data in a pipeline of trasformations" do
    actual = TestSupport.log_test_log_r3()
    as_expected? = Regex.match?(~r/[first: "john"]/, actual) &&
                              Regex.match?(~r/module: Elixir.Og.TestSupport/, actual) &&
                              Regex.match?(~r/function: log_test_log_r3\/0/, actual) &&
                              Regex.match?(~r/line: /, actual)
    assert(as_expected? == :true)
  end

  test "log_r returns intermediate data unchanged in a pipeline of trasformations" do
    Logger.configure([level: :error]) # ignore logs for this test below log_level :error
    actual =
      %{first: "john", last: "doe"}
      |> Map.to_list()
      |> Enum.filter( &(&1 === {:first, "john"}))
      |> Og.log_r()
      |> List.last()
      |> Tuple.to_list()
      |> List.last()
      |> String.upcase()
    Logger.configure([level: :debug]) # change back the log_level
    expected = "JOHN"
    assert(actual ==expected)
  end

  test "log sanitizes input as expected (1)" do
    actual = CaptureLog.capture_log(
      [level: :debug],
      fn() -> Og.log(%{credit_card: "4111111111111111"}, sanitize: :true, level: :info) end
    )
    as_expected? = Regex.match?(~r/HIDDEN/, actual)
    assert(as_expected? == :true)
  end

  test "log sanitizes input as expected (2)" do
    actual = CaptureLog.capture_log(
      [level: :debug],
      fn() -> Og.log("Credit_card number: 4111111111111111", sanitize: :true, level: :error) end
    )
    as_expected? = Regex.match?(~r/HIDDEN/, actual)
    assert(as_expected? == :true)
  end

  # The expected output should not be printed since the `:compile_time_purge_level` is set to `:info`
  test "compile_time_purge_level works as expected - :debug" do
    actual = TestSupport.compile_purge_test(:debug)
    as_expected? = Regex.match?(~r/test/, actual)
    refute(as_expected? == :true)
  end

  # The expected output should be printed since the `:compile_time_purge_level` is set to `:info`
  test "compile_time_purge_level works as expected - :info" do
    actual = TestSupport.compile_purge_test(:info)
    as_expected? = Regex.match?(~r/test/, actual)
    assert(as_expected? == :true)
  end

end


test/support/test_support.ex
defmodule Og.TestSupport do
  alias ExUnit.CaptureLog
  @data "test"


  def log_test(capture_level, log_level) do
    CaptureLog.capture_log(
      [level: capture_level],
      fn() -> Og.log(@data, env: __ENV__, level: log_level) end
    )
  end

  def log_test_log_r1() do
    CaptureLog.capture_log(
      [level: :debug],
      fn() ->
        map = %{first: "john", last: "doe"}
        Map.to_list(map)
        |> Enum.filter( &(&1 ==={:first, "john"}))
        |> Og.log_r(level: :info)
        |> List.last()
        |> Tuple.to_list()
        |> List.last()
        |> String.upcase()
      end
    )
    #  [first: "john"]
  end


  def log_test_log_r2() do
    CaptureLog.capture_log(
      [level: :debug],
      fn() ->
        map = %{first: "john", last: "doe"}
        Map.to_list(map)
        |> Enum.filter( &(&1 ==={:first, "john"}))
        |> Og.log_r(env: __ENV__, level: :info)
        |> List.last()
        |> Tuple.to_list()
        |> List.last()
        |> String.upcase()
      end
    )
    #  module: ..., function: ..., line: ..., [first: "john"]
  end


  def log_test_log_r3() do
    CaptureLog.capture_log(
      [level: :debug],
      fn() ->
        map = %{first: "john", last: "doe"}
        Map.to_list(map)
        |> Enum.filter( &(&1 ==={:first, "john"}))
        |> Og.log_r(env: __ENV__, level: :info)
        |> List.last()
        |> Tuple.to_list()
        |> List.last()
        |> String.upcase()
      end
    )
  end

  def compile_purge_test(:debug) do
    CaptureLog.capture_log(
      [level: :debug],
      fn() -> Og.log(@data, level: :debug) end
    )
  end

  def compile_purge_test(:info) do
    CaptureLog.capture_log(
      [level: :debug],
      fn() -> Og.log(@data, level: :info) end
    )
  end

end

mix.exs
defmodule Og.Mixfile do
  use Mix.Project
  @version "1.1.1"

  def project do
    [
     app: :og,
     name: "Óg",
     version: @version,
     elixir: ">= 1.2.0",
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     elixirc_paths: elixirc_paths(Mix.env),
     source_url: "https://github.com/stephenmoloney/og",
     description: "Og is a small collection of debugging functions for use during development.",
     package: package(),
     deps: deps(),
     docs: [
       main: "Og",
       extra_section: ""
     ]
    ]
  end

  def application do
    [applications: [:logger, :apex, :secure_log_formatter]]
  end

  def deps() do
    [
      {:apex,  "~> 1.0"},
      {:secure_log_formatter, "~> 1.1"},
      {:earmark, "~> 1.0", only: :dev},
      {:ex_doc,  "~> 0.14", only: :dev}
    ]
  end

  defp package do
    %{
      licenses: ["MIT"],
      maintainers: ["Stephen Moloney"],
      links: %{ "GitHub" => "https://github.com/stephenmoloney/og"},
      files: ~w(lib mix.exs README* LICENSE* CHANGELOG*)
     }
  end


  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

end


config/dev.exs
use Mix.Config

config :logger,
  backends: [:console],
  level: :debug,
  compile_time_purge_level: :debug,
  compile_time_application: :my_app,
  truncate: (4096 * 8),
  utc_log: :false

config :logger, :console,
  level: :debug,
  format: "$time $metadata [$level] $message\n",
  metadata: [:module, :function, :line]

config :logger, :og,
  kernel_opts: [width: 70],
  apex_opts: [numbers: :false, color: :false],
  sanitize_by_default: :false,
  default_inspector: :kernel


config :logger,
  secure_log_formatter:
    [
      fields: [~r/\w*_token/, "credit_card", "password"],
      patterns: [
        ~r/#?(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})/,
        ~r/\d{3}-?\d{2}-?\d{4}/
      ],
      replacement: "[ERASED]"
    ]

lib/og.ex
defmodule Og do
  @moduledoc ~S"""
  Óg is a small collection of debugging helper functions. Og is a debugging tool primarily
  intendend for development, the use of ordinary `Logger` is preferred for production.

  ## Summary

  - `log/2` - logs the data transformed by the inspector function
  and returns `:ok`


  - `log_r/2` - logs the data transformed by the inspector function
  and returns the original data.


  - Inspection of the data before logging it can be helpful in a debugging context for
  example for
      - Avoiding the `Protocol.UndefinedError` when logging tuples.
      - Not needing to require Logger


  - Formatting with `Kernel.inspect` on the data carries an overhead so
  `Og.log/2` and `Og.log_r/2` should be probably be reserved for debugging code in `:dev`
  environments.


  ## Configuration

      config :logger, :og,
        kernel_opts: [width: 70],
        apex_opts: [numbers: :false, color: :false],
        sanitize_by_default: :false,
        default_inspector: :kernel


  ## Configuration options

  - `kernel_opts` - corresponds to elixir inspect opts for `IO.inspect/2`
  and `Kernel.inspect/2`, refer to `https://hexdocs.pm/elixir/Inspect.Opts.html`


  - `apex_opts` - corresponds to options for the `Apex.Format.format/2` function,
  refer to `https://github.com/BjRo/apex/blob/master/lib/apex/format.ex`


  - `sanitize_by_default` - defaults to `:false`, when set to `:true`, an attempt will
  be made to apply the `SecureLogFormatter.sanitize/1` function on the data
  before applying the inspection function. For this function to take any
  effect, the settings for `SecureLogFormatter` must also be placed in
  `config.exs`. See
  [secure_log_formatter](https://github.com/localvore-today/secure_log_formatter/blob/master/lib/secure_log_formatter.ex)
  for more details.


  - `default_inspector` - corresponds to the default inspector which will apply
  to the data passed to the log function. This can be overriden in the options
  of the log function. The options are `:kernel` or `:apex`, the default is `:kernel`.


  Example configuration for secure_log_formatter, as referenced at
  the following [secure_log_formatter url](https://github.com/localvore-today/secure_log_formatter/blob/master/lib/secure_log_formatter.ex).


      config :logger,
      secure_log_formatter:
        [
          # Map and Keyword List keys who's value should be hidden
          fields: ["password", "credit_card", ~r/.*_token/],

          # Patterns which if found, should be hidden
          patterns: [~r/4[0-9]{15}/] # Simple credit card example

          # defaults to "[REDACTED]"
          replacement: "[PRIVATE]"
        ]
  """
  require Logger
  defp kernel_opts(), do: Application.get_env(:logger, :og, [])
                          |> Keyword.get(:kernel_opts, [])
  defp apex_opts(), do: Application.get_env(:logger, :og, [])
                        |> Keyword.get(:apex_opts, [color: :false, numbers: :false])
  defp default_inspector(), do: Application.get_env(:logger, :og, [])
                        |> Keyword.get(:default_inspector, :kernel)
  defp sanitize_by_default?(), do: Application.get_env(:logger, :og, [])
                        |> Keyword.get(:sanitize_by_default, :false)


  # Public

  @doc """
  Formats the data using an inspector function, logs it and returns the
  atom `:ok`.

  ## Notes:

  There is an overhead in converting the data to
  other formats such as a binary representation. Hence,
  `Og.log/2` and `Og.log_r/2` are preferred for
  development debugging purposes only.

  ## opts

  - ***level***: defaults to `:debug`.

  - ***env***: defaults to `:nil`.

  - ***inspector***: defaults to `:default_inspector` in the application config.exs and if not
  set falls back to `Kernel.inspect/2`.
  The inspector function determines what function transforms the data prior to logging.
  Currently the options are `:kernel` or `:apex` which use
  the functions `&Kernel.inspect/2` and `&Apex.Format.format/2` respectively.

  - ***sanitize***: defaults to `:sanitize_by_default` in the application config.exs and if not
  set falls back to `:false`.
  When set to `:true` the function `SecureLogFormatter.sanitize/1` will be applied to
  the data prior to logging the data. See
  [secure_log_formatter](https://github.com/localvore-today/secure_log_formatter/blob/master/lib/secure_log_formatter.ex)
  for more details.

  ## Examples:

      Og.log(%{test: "test"})

      Og.log(%{test: "test"}, level: :info)

      Og.log(%{test: "test"}, env: __ENV__)

      Og.log(%{test: "test"}, inspector: :apex)

      Og.log(%{credit_card: "4111111111111111"}, sanitize: :true)
  """
  @spec log(any, Keyword.t) :: :ok
  def log(data, opts \\ [])
  def log(data, opts) when is_list(opts) do
    inspector = Keyword.get(opts, :inspector, default_inspector())
    sanitize? = Keyword.get(opts, :sanitize, sanitize_by_default?())
    inspector_opts =
    case inspector do
      :kernel -> kernel_opts()
      :apex -> apex_opts()
      _ -> kernel_opts()
    end
    inspector =
    case inspector do
      :kernel -> &Kernel.inspect/2
      :apex -> &Apex.Format.format/2
      _ -> &Kernel.inspect/2
    end
    data =
    case sanitize? do
      :true -> SecureLogFormatter.sanitize(data)
      :false -> data
    end
    env = Keyword.get(opts, :env, :nil)
    level = Keyword.get(opts, :level, :debug)
    (base_details(env) <> inspector.(data, inspector_opts))
    |> log_data(level)
    :ok
  end

  @spec log(data :: any, env :: Macro.Env.t) :: :ok
  def log(data, env_or_level)
  @spec log(data :: any, level :: atom) :: :ok
  def log(data, :error), do: Og.log(data, level: :error)
  def log(data, :warn), do: Og.log(data, level: :warn)
  def log(data, :info), do: Og.log(data, level: :info)
  def log(data, :debug), do: Og.log(data, level: :debug)

  @spec log(data :: any, env :: Macro.Env.t, level :: atom) :: :ok
  def log(data, %Macro.Env{} = arg, :error), do: Og.log(data, env: arg, level: :error)
  def log(data, %Macro.Env{} = arg, :warn), do: Og.log(data, env: arg, level: :warn)
  def log(data, %Macro.Env{} = arg, :info), do: Og.log(data, env: arg, level: :info)
  def log(data, %Macro.Env{} = arg, :debug), do: Og.log(data, env: arg, level: :debug)
  @spec log(data :: any, level :: atom, env :: Macro.Env.t) :: :ok
  def log(data, :error, %Macro.Env{} = arg), do: Og.log(data, env: arg, level: :error)
  def log(data, :warn, %Macro.Env{} = arg), do: Og.log(data, env: arg, level: :warn)
  def log(data, :info, %Macro.Env{} = arg), do: Og.log(data, env: arg, level: :info)
  def log(data, :debug, %Macro.Env{} = arg), do: Og.log(data, env: arg, level: :debug)


  @doc """
  Formats the data using an inspector function, logs it and returns the
  original data.

  ## Notes:

  There is an overhead in converting the data to
  other formats such as a binary representation. Hence,
  `Og.log/2` and `Og.log_r/2` are preferred for
  development debugging purposes only.

  ## opts

  - ***level***: defaults to `:debug`.

  - ***env***: defaults to `:nil`.

  - ***inspector***: defaults to `:default_inspector` in the application config.exs.
  The inspector function determines what function transforms the data prior to logging.
  Currently the options are `:kernel` or `:apex` which use
  the functions `&Kernel.inspect/2` and `&Apex.Format.format/2` respectively.

  - ***sanitize***: defaults to `:sanitize_by_default` in the application config.exs.
  When set to `:true` the function `SecureLogFormatter.sanitize/1` will be applied to
  the data prior to logging the data. See
  [secure_log_formatter](https://github.com/localvore-today/secure_log_formatter/blob/master/lib/secure_log_formatter.ex)
  for more details.


  ## Examples:

      Og.log(%{test: "test"})

      Og.log(%{test: "test"}, level: :info)

      Og.log(%{test: "test"}, env: __ENV__)

      Og.log(%{test: "test"}, inspector: :apex)

      Og.log(%{credit_card: "4111111111111111"}, sanitize: :true)
  """
  @spec log_r(any, Keyword.t) :: any
  def log_r(data, opts \\ [])
  def log_r(data, opts) when is_list(opts) do
    log(data, opts)
    data
  end

  @spec log_r(data :: any, env :: Macro.Env.t) :: any
  def log_r(data, %Macro.Env{} = env), do: Og.log_r(data, env: env)
  @spec log_r(data :: any, level :: atom) :: any
  def log_r(data, :error), do: Og.log_r(data, level: :error)
  def log_r(data, :warn), do: Og.log_r(data, level: :error)
  def log_r(data, :info), do: Og.log_r(data, level: :info)
  def log_r(data, :debug), do: Og.log_r(data, level: :debug)

  @spec log_r(data :: any, env :: Macro.Env.t, level :: atom) :: any
  def log_r(data, %Macro.Env{} = arg, :error), do: Og.log_r(data, env: arg, level: :error)
  def log_r(data, %Macro.Env{} = arg, :warn), do: Og.log_r(data, env: arg, level: :warn)
  def log_r(data, %Macro.Env{} = arg, :info), do: Og.log_r(data, env: arg, level: :info)
  def log_r(data, %Macro.Env{} = arg, :debug), do: Og.log_r(data, env: arg, level: :debug)
  @spec log_r(data :: any, level :: atom, env :: Macro.Env.t) :: any
  def log_r(data, :error, %Macro.Env{} = arg), do: Og.log_r(data, env: arg, level: :error)
  def log_r(data, :warn, %Macro.Env{} = arg), do: Og.log_r(data, env: arg, level: :warn)
  def log_r(data, :info, %Macro.Env{} = arg), do: Og.log_r(data, env: arg, level: :info)
  def log_r(data, :debug, %Macro.Env{} = arg), do: Og.log_r(data, env: arg, level: :debug)


  # Private


  defp base_details(:nil), do: ""
  defp base_details(env) do
    put_into_string = fn(elem, name) ->
      case elem do
        :nil -> :nil
        _ -> "#{name}: #{elem}, "
      end
    end
    function =
    case env.function do
      :nil ->  :nil
      {function, arity} -> "#{function}/#{arity}"
    end
    application = Application.get_env(:logger, :compile_time_application) |> put_into_string.("application")
    module = put_into_string.(env.module, "module")
    function = put_into_string.(function, "function")
    line = put_into_string.(env.line, "line") |> String.trim_trailing(" ") |> String.trim_trailing(",")
    "#{application}#{module}#{function}#{line}:\n"
  end


  defp log_data(formatted_data, level) do
    cond do
      level == :error -> Logger.error(formatted_data)
      level == :warn -> Logger.warn(formatted_data)
      level == :info -> Logger.info(formatted_data)
      level == :debug -> Logger.debug(formatted_data)
      :true -> raise("unexpected log level")
    end
  end

end
